src/core/net/ip6.cpp

```diff
@@ -1199,9 +1236,16 @@ otError Ip6::HandleDatagram(Message &aMessage, Netif *aNetif, const void *aLinkM
         }
 #endif
 
-        IgnoreError(ProcessReceiveCallback(aMessage, messageInfo, nextHeader, aFromNcpHost, Message::kCopyToUse));
+        error = ProcessReceiveCallback(aMessage, messageInfo, nextHeader, aFromNcpHost, Message::kCopyToUse);
+
+        if ((error == OT_ERROR_NONE || error == OT_ERROR_NO_ROUTE) && forward &&
+            !ShouldForwardToThread(messageInfo, aFromNcpHost))
+        {
+            forward = false;
+        }
```

Comment:
The `if()` block above seems to be new logic/code.  
@bukepo can you please help explain what it does (why we need it)? Thanks.

Reply:
This prevents `ProcessReceiveCallback` be called multiple times. When line 1239 succeeds, I don't think the host needs to receive the message again. And if it fails for no route, another trial doesn't help. However this doesn't prevent trying again for failures caused by insufficient memory.